plugins/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpProcessor.java

```diff
@@ -352,33 +372,52 @@ public GeoIpProcessor create(Map<String, Processor.Factory> registry, String pro
                     "database_file", "database file [" + databaseFile + "] doesn't exist");
             }
 
-            DatabaseReader databaseReader = lazyLoader.get();
-            String databaseType = databaseReader.getMetadata().getDatabaseType();
+            final Supplier<DatabaseReader> databaseReader = () -> {
+                try {
+                    return lazyLoader.get();
+                } catch (final IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            };
 
-            final Set<Property> properties;
+            final Supplier<Set<Property>> propertiesSupplier;
             if (propertyNames != null) {
-                properties = EnumSet.noneOf(Property.class);
-                for (String fieldName : propertyNames) {
-                    try {
-                        properties.add(Property.parseProperty(databaseType, fieldName));
-                    } catch (IllegalArgumentException e) {
-                        throw newConfigurationException(TYPE, processorTag, "properties", e.getMessage());
+                final AtomicBoolean set = new AtomicBoolean();
+                final Set<Property> properties = EnumSet.noneOf(Property.class);
+                propertiesSupplier = () -> {
+                    if (set.compareAndSet(false, true)) {
```

Comment:
I think this publishes `properties` racily? E.g. one thread could enter the block guarded by the `AtomicBoolean` and is adding properties to the set in the loop while another one could read `properties` which is not fully initialized yet? Also, I am not aware that there is a happens-before relation between `EnumSet#add` and `EnumSet#iterator()`?

Reply:
You're right but it ultimately doesn't matter. This code has to be scrapped so that we can maintain eager configuration validation. I'm still looking for a clean way to do this without losing the lazy loading.